node/zombienet-backchannel/src/lib.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2017-2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Provides the possibility to coordination between malicious actors and
+//! the zombienet test-runner, allowing to reference runtime's generated
+//! values in the test specifications, through a bidirectional message passing
+//! implemented as a `backchannel`.
+
+use futures_util::{stream::SplitSink, SinkExt, StreamExt};
+use parity_scale_codec as codec;
+use serde::{Deserialize, Serialize};
+use std::env;
+use tokio::{net::TcpStream, sync::broadcast};
+use tokio_tungstenite::{
+	connect_async, tungstenite::protocol::Message, MaybeTlsStream, WebSocketStream,
+};
+
+mod errors;
+use errors::BackchannelError;
+
+#[derive(Debug)]
+pub struct ZombienetBackchannel {
+	broadcast_tx: broadcast::Sender<BackchannelItem>,
+	ws_tx: SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct BackchannelItem {
+	key: String,
+	value: String,
+}
+
+pub const ZOMBIENET: &str = "ZOMBIENET_BACKCHANNEL🧟🧟🧟";
+
+impl ZombienetBackchannel {
+	pub async fn init() -> Result<ZombienetBackchannel, BackchannelError> {
```

Comment:
I think this should be split into `fn init() -> Result<(), _> {}` and fn sender() -> Result<ZombienetBackchannel, BackchannelError>`

Reply:
Hi @drahnr, thanks for the feedback! Just to clarify:
 
- The idea is to call `init` at the start of the `malus`, and that will create the `ws` connection to the `zombienet` test-runner.

```rust
zombienet::init().await?
```

Then in other places we want to have access to either `send` or `subscribe` using the `broadcaster` associated fn and return an instance to allow us to interact with the underline `ws` like
```rust
zombienet::broadcaster().send(Foo(Something))
```
or
```rust
let receiver = zombienet::broadcaster().subscribe();
```

So broadcaster will need access to the initially created instance. I can implement this as a singleton if it's ok with you.

Thanks!

Reply:
Exactly :) Singelton is fine, it's a test helper only.